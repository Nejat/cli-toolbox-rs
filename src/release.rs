/// The `release!` macro provides release only expression and statement evaluation
#[macro_export]
macro_rules! release {
    // ===============================================================================
    // evaluate expression in release
    ($expr:expr) => {
        #[cfg(not(debug_assertions))]
        $expr
    };
    // ===============================================================================
    // evaluate expression in release if verbosity is quite
    (QUITE $expr:expr) => {
        #[cfg(not(debug_assertions))]
        if Verbosity::is_quite() { $expr }
    };
    // evaluate expression in release if verbosity is terse
    (TERSE $expr:expr) => {
        #[cfg(not(debug_assertions))]
        if Verbosity::is_terse() { $expr }
    };
    // evaluate expression in release if verbosity is verbose
    (VERBOSE $expr:expr) => {
        #[cfg(not(debug_assertions))]
        if Verbosity::is_verbose() { $expr }
    };
    // ===============================================================================
    // evaluate terse or verbose expression in release if verbosity matches
    (
        TERSE   $terse_expr:expr;
        VERBOSE $verbose_expr:expr;
    ) => {
        #[cfg(not(debug_assertions))]
        match Verbosity::level() {
            Verbosity::Terse   => { $terse_expr }
            Verbosity::Verbose => { $verbose_expr }
            _ => {}
        }
    };
    // ===============================================================================
    // evaluate statement in release
    ($stmt:stmt) => {
        #[cfg(not(debug_assertions))]
        $stmt
    };
    // ===============================================================================
    // evaluate statement in release if verbosity is quite
    (QUITE => $stmt:stmt) => {
        #[cfg(not(debug_assertions))]
        if Verbosity::is_quite() { $stmt }
    };
    // evaluate statement in release if verbosity is terse
    (TERSE => $stmt:stmt) => {
        #[cfg(not(debug_assertions))]
        if Verbosity::is_terse() { $stmt }
    };
    // evaluate statement in release if verbosity is verbose
    (VERBOSE => $stmt:stmt) => {
        #[cfg(not(debug_assertions))]
        if Verbosity::is_verbose() { $stmt }
    };
    // ===============================================================================
    // evaluate terse or verbose statement in release if verbosity matches
    (
        TERSE   => $terse_expr:stmt,
        VERBOSE => $verbose_expr:stmt
    ) => {
        #[cfg(not(debug_assertions))]
        match Verbosity::level() {
            Verbosity::Terse   => { $terse_expr }
            Verbosity::Verbose => { $verbose_expr }
            _ => {}
        }
    };
}
