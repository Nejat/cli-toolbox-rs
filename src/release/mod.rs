use syn::Expr;
use verbosity::Verbosity;

mod parse;
mod tokenize;

pub struct Release {
    expr: Expr,
    verbosity: Verbosity
}