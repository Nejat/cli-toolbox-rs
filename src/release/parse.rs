use syn::parse::{Parse, ParseStream};

use crate::common::parse::{parse_expression, parse_verbosity};
use crate::release::Release;

impl Parse for Release {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(Self {
            verbosity: parse_verbosity(input)?,
            expr: parse_expression(input, "release")?,
        })
    }
}