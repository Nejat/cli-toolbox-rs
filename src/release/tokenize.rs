use proc_macro2::TokenStream;
use quote::ToTokens;
use crate::common::tokenize::tokenize_verbosity_checked_expr;

use crate::release::Release;

impl ToTokens for Release {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let expr = tokenize_verbosity_checked_expr(self.verbosity, &self.expr);

        tokens.extend(
            quote! {
                #[cfg(not(debug_assertions))]
                #expr
            }
        );
    }
}