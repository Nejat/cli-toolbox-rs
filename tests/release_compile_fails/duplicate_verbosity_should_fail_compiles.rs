use cli_toolbox::release;

fn main() {
    release! {
        @terse println!("terse message: {}", 42);
        @terse println!("terse message: {}", 42)
    }

    release! {
        @verbose println!("verbose message: {}", 42);
        @verbose println!("verbose message: {}", 42)
    }
}