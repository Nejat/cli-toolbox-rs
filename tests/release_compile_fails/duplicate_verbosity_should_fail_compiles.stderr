error: do not duplicate verbosity
 --> tests/release_compile_fails/duplicate_verbosity_should_fail_compiles.rs:6:9
  |
6 |         @terse println!("terse message: {}", 42)
  |         ^

error: do not duplicate verbosity
  --> tests/release_compile_fails/duplicate_verbosity_should_fail_compiles.rs:11:9
   |
11 |         @verbose println!("verbose message: {}", 42)
   |         ^
